### ISSUE TITLE
mixer: tune/adjust the number of goroutines + goroutine leak (in tests)

### ISSUE THREAD

**Issue by ldemailly** [istio/istio#3948]
It appears the default setting in mixer has 2x 1024 goroutine pools
yet fail some test by exceeding an 8192 limit

a) find/fix the leak
b) set the number lower for tests
c) provide guidelines for prod/scale tuning on what the number should be (why 1024,...)

See the discussion in https://github.com/istio/istio/pull/3795

**Comment by ozevren** [comment]
This issue is misleading. We're using mixer server in a similar fashion in a bunch of other tests without failure. The issue seems to be isolated to that particular test (so far) which is already tracked by the previous issue.

If you have a particular reasoning for the current defaults being wrong, please add it to the bug and we can discuss. Otherwise this bug should really be about item c.

We should try to open bugs on focused problems & deliverables. If you think there are more issues here other than c, let's  track them independently.

**Comment by ldemailly** [comment]
yes I do think there are 3 issues to be addressed:

for a) a 8192 limit should not trigger and removing the test or the limit isn't how that should be addressed

for b) it is inefficient and we want the test to be as fast as possible

for c) it's important for our users to have some guidelines and reasoning behind numbers

**Comment by ozevren** [comment]
Agreed on a), but that really should be addressed as part of that test and
is not a general Mixer problem. The previous bug should be capturing this.

Similarly for b, I am not sure what you mean is inefficient? Changing the
number of goroutines in the pool will not affect test runtimes in a
meaningful way.

So, the real issue seems to be c), and we should really convert this bug to
cover that.

**Comment by ldemailly** [comment]
I'll split into 3 issues if you feel that is necessary

**Comment by ozevren** [comment]
My main point was not to split this into 3 things. My main points are:
- This is a particular issue with the test in question only. If we fix the test (i.e. make it stop leaking), this should go away. This issue is already being tracked by the original bug, so let's use that to fix the issue correctly, and create bugs for the things cannot be fixed within that context.
- Changing the default values from 1024 to something else neither helps the tests to run faster, nor helps with the leak. At best, it papers over the issue. 
- Of all the points discussed, the only actionable item is documenting these flags, providing rationale for the current values of these flags. We should have 1 bug (ideally this one repurposed) specifically focused on that.

**Comment by ldemailly** [comment]
Are you confident the code test is leaking the go routines and it's related exclusively to the test and not to mixer goroutine pool ? 

Maybe I mis interpreted - when I asked where in our system could we be having thousands of goroutine, the answer I was given was mixer... if that's not correct or the thousands in mixer are unrelated to the test completely then yes we can completely decouple a) from b) and c)

In case it's not obvious, I never advocate papering over bugs - there is a reason for the order; obviously we should understand and fix the 8k limit being hit _before_ we lower the limit. yet once we solved that problem, lowering resources consumption during basic tests is just natural (b)) and getting data about what a good number is is as well (c))

**Comment by ozevren** [comment]
Yes, based on my discussions with Joy and my observations of already
existing tests I am pretty confident. Of course, being a fallible human I
may be wrong. Trying to fix the test by doing proper cleanup should give us
the right information. If we find that there is indeed a deeper problem, we
can investigate further.
